mlir/lib/Catalyst/Transforms/InlineNestedModules.cpp

```diff
@@ -466,6 +537,31 @@ struct InlineNestedSymbolTablePass : PassWrapper<InlineNestedSymbolTablePass, Op
             signalPassFailure();
         }
 
+        // Restore external API func decl names and update calls.
+        mlir::DenseMap<StringRef, StringAttr> uniqued_names_to_APIName;
+        symbolTable->walk([&](func::FuncOp funcOp) {
+            if (funcOp->hasAttr("original_external_API_name")) {
+                uniqued_names_to_APIName[funcOp.getSymName()] =
+                    cast<StringAttr>(funcOp->getAttr("original_external_API_name"));
+            }
+        });
+        RewritePatternSet restoreExternalFuncDeclName(context);
+        restoreExternalFuncDeclName.add<RestoreExternalFuncDeclNamePattern>(context);
+        run = _stopAfterStep >= 5 || _stopAfterStep == 0;
+        if (run && failed(applyPatternsGreedily(symbolTable, std::move(restoreExternalFuncDeclName),
+                                                config))) {
+            signalPassFailure();
+        }
+
+        RewritePatternSet updateCalleeToExternalAPINames(context);
+        updateCalleeToExternalAPINames.add<UpdateCalleeToExternalAPINamesPattern>(
+            context, uniqued_names_to_APIName);
+        run = _stopAfterStep >= 5 || _stopAfterStep == 0;
+        if (run && failed(applyPatternsGreedily(
+                       symbolTable, std::move(updateCalleeToExternalAPINames), config))) {
+            signalPassFailure();
+        }
+
```

Comment:
Wow there is a lot of invocations of the greedy pattern applicator in this pass. I think this may indicate a misuse of the infrastructure.
Patterns are meant to be small rewrites that are independent, composable, and repeatable (typically).
If this is not the case for a pass, and instead we want to apply a _specific sequence_ of rewrites _one time_, then we should do so directly which will be much more efficient.

I'm not saying this is 100% the case here, but it definitely looks suspicious.